Fixed formatting issue caused by tryAddInfoRightLine().

swstegall opened the pull request:
Issue #3633 demonstrates that formatting issues in the profile will occur if a supplied string in the profile header is too large. After testing, I found that supplying tryAddInfoRightLine() with a string with more than 34 characters will cause formatting to break. To replicate how this is handled on the osu.Web interface, I added a case where any string with over 34 characters supplied to tryAddInfoRightLine() will print a substring of the first 31 characters and an ellipsis to the string before passing it along to infoTextRight.AddText().

Add any details pertaining to developers above the break.

- [ ] Depends on #PR
- Closes #3633

---

Fixed formatting issue in User ProfileHeader with strings with more than 34 characters.

Add a sentence or two describing this change in plain english. This will be displayed on the [changelog](https://osu.ppy.sh/home/changelog). A single screenshot or short gif is also welcomed.

swstegall committed:
Fixed formatting issue caused by tryAddInfoRightLine().

Issue #3633 demonstrates that formatting issues in the profile will occur if a supplied string in the profile header is too large. After testing, I found that supplying tryAddInfoRightLine() with a string with more than 34 characters will cause formatting to break. To replicate how this is handled on the osu.Web interface, I added a case where any string with over 34 characters supplied to tryAddInfoRightLine() will print a substring of the first 31 characters and an ellipsis to the string before passing it along to infoTextRight.AddText().

jorolf commented:
You don't have to create 3 different pull requests. You should've added changes to the first one.

swstegall commented:
> You don't have to create 3 different pull requests. You should've added changes to the first one.

That was my mistake. I'm still fairly new to GitHub and I'm sure my excessive pull requests do not look good.

I still do not understand why my build checks are failing. AppVeyor complains about excessive whitespace on a certain line, but that line contains no additional whitespace, hence why I made the improper choice of duplicating my commit with changes to try and rectify this issue.

morguldir commented on osu.Game/Overlays/Profile/ProfileHeader.cs:
```diff
@@ -462,16 +462,23 @@ private void loadUser()
 
         private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
         {
+            const int LocalCharLimit = 31; 
```
```suggestion
            const int LocalCharLimit = 31;
```

morguldir committed:
Fixed whitespace issue.

Co-Authored-By: swstegall <swstegall@hotmail.com>

swstegall committed:
Fixed update to conform to consistent programming style.

Removed two cases of extraneous parenthesis and one improperly-named variable.

VPeruS commented on osu.Game/Overlays/Profile/ProfileHeader.cs:
```diff
@@ -462,16 +462,23 @@ private void loadUser()
+            const int local_char_limit = 31;
             if (string.IsNullOrEmpty(str)) return;
 
             infoTextRight.AddIcon(icon);
```
```suggestion
             infoTextRight.AddIcon(icon);

             var line = " " + str;
             if (line.Length > local_char_limit)
                 line = str.Substring(0, local_char_limit) + "..."
```

Aergwyn commented:
This seems to be about the good ol' truncating topic? See https://github.com/ppy/osu/pull/2826 for reference where something similar was discussed.

VPeruS commented on osu.Game/Overlays/Profile/ProfileHeader.cs:
```diff
@@ -462,16 +462,23 @@ private void loadUser()
 
         private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
         {
+            const int local_char_limit = 31;
```
Glyphs can have different width, for example size of "i" != "w".

swstegall committed:
Merge branch 'master' into patch-1

swstegall commented:
> This seems to be about the good ol' truncating topic? See #2826 for reference where something similar was discussed.

After analyzing the suggestions in that pull request, I still find myself a bit lost on how this should be handled. Are you suggesting passing the truncation off to another method and then including that method inside tryAddInfoRightLine()? Perhaps something like this:
```csharp
private string Truncate(string str, int bound)
{
    if (str.Length > bound)
        return str.Substring(0, bound) + "...");
}
```

And then calling that method inside tryAddInfoRightLine() with a block like this:
```csharp
private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
{
    const int local_char_limit = 31;
    if (string.IsNullOrEmpty(str) == true) return;
    
    infoTextRight.AddIcon(icon);

    var line = " " + str;
    if (line.Length > local_char_limit)
        Truncate(line, local_char_limit);
    if(url != null)
        infoTextRight.AddLink(line, url);
    else
        infoTextRight.AddText(line);

    infoTextRight.NewLine();
}
```

swstegall committed:
Abstracted truncation to private method.

Abstracted truncation of long strings that are used in tryAddInfoRightLine() to a method named Truncate() to simplify the readability of tryAddInfoRightLine().

swstegall committed:
Removed tab characters.

Attempted to remove tabs caused by my configuration for vim.

Aergwyn commented:
No. It was more about that a truncate logic should exist at `SpriteText` level so this kind of localised logic that we'd need at many more places isn't necessary.

swstegall committed:
Removed tab characters.

Attempted a different method to remove tab characters. I failed to set the correct property in vim in my previous commit to handle these properly, but I'm fairly sure that the method that I used in this attempt solves this issue.

swstegall committed:
Renamed method to conform to suggested style.

swstegall commented:
> No. It was more about that a truncate logic should exist at `SpriteText` level so this kind of localised logic that we'd need at many more places isn't necessary.

I see. I will have to do some more research and testing before attempting to develop a solution that would work on that scale. Thank you for your patience.

smoogipoo reviewed (changes requested):
* Truncation should be implemented inside SpriteText itself (osu!framework).
* You changed the line ends of the entire file.

Joehuu commented:
Can be closed as referenced issue in OP was resolved with new design.
